Validator doesn't resolve refs for a whole response

**Steps to reproduce:**

1. Take an OAS3 file like:
```openapi: 3.0.0
info:
  title: example
  version: "0.1"
  description: test
servers:
  - url: /api
    description: local
paths:
  /feature:
    post:
      description: Create new feature in a system
      responses:
        "400":
          $ref: '#/components/responses/InvalidRequest'
components:
  responses:
    InvalidRequest:
      description: Request was rejected by server due to invalid request parameters or something
      content:
        application/json:
          schema:
            $ref: '#/components/schemas/Error'
```

2. Use it as the specification for a test
3. Let the test check POST /api/feature with response "404"

Actual result: 
```
AssertionError: expected response to satisfy API spec:
{ message: 'The response was not valid.',
  errors:
   [ { errorCode: 'type.openapi.responseValidation',
       message: 'response should be null' } ],
  actualResponse: { status: 400, body: { error: [Object] } } }```

--- a/lib/index.js
+++ b/lib/index.js
@@ -94,7 +94,11 @@ function findResExpectedByApiSpec(res, openApiSpec) {
   }
 
   const status = res.status();
-  const expectedResponse = endpointObj.responses[status];
+  let expectedResponse = endpointObj.responses[status];
+  if (expectedResponse && expectedResponse['$ref']) {
+    const responseSchemaName = expectedResponse['$ref'].split('#/components/responses/')[1];
+    expectedResponse = openApiSpec.components.responses[responseSchemaName];
+  }
   if (!expectedResponse) {
     throw new Error(`No '${status}' response defined for endpoint '${httpMethod} ${openApiPath}' in OpenAPI spec`);
   }

--- a/lib/index.js
+++ b/lib/index.js
@@ -66,8 +66,7 @@ function parseOpenApiFile(filePath) {
   }
 
   try {
-    const openApiVersion = openApiSpec.swagger || openApiSpec.openapi;
-    const validator = new OpenAPISchemaValidator({ version: openApiVersion });
+    const validator = new OpenAPISchemaValidator({ version: getOAVersion(openApiSpec) });
     const { errors } = validator.validate(openApiSpec);
     if (errors.length > 0) {
       throw new Error(utils.stringify(errors));
@@ -77,6 +76,10 @@ function parseOpenApiFile(filePath) {
   }
 
   return openApiSpec;
+}
+
+function getOAVersion(openApiSpec) {
+  return openApiSpec.swagger || openApiSpec.openapi;
 }
 
 function findResExpectedByApiSpec(res, openApiSpec) {
@@ -96,8 +99,8 @@ function findResExpectedByApiSpec(res, openApiSpec) {
   const status = res.status();
   let expectedResponse = endpointObj.responses[status];
   if (expectedResponse && expectedResponse['$ref']) {
-    const responseSchemaName = expectedResponse['$ref'].split('#/components/responses/')[1];
-    expectedResponse = openApiSpec.components.responses[responseSchemaName];
+    const nameOfReferencedResponseDefinition = expectedResponse['$ref'].split('#/components/responses/')[1];
+    expectedResponse = openApiSpec.components.responses[nameOfReferencedResponseDefinition];
   }
   if (!expectedResponse) {
     throw new Error(`No '${status}' response defined for endpoint '${httpMethod} ${openApiPath}' in OpenAPI spec`);

--- a/lib/index.js
+++ b/lib/index.js
@@ -99,14 +99,23 @@ function findResExpectedByApiSpec(res, openApiSpec) {
   const status = res.status();
   let expectedResponse = endpointObj.responses[status];
   if (expectedResponse && expectedResponse['$ref']) {
-    const nameOfReferencedResponseDefinition = expectedResponse['$ref'].split('#/components/responses/')[1];
-    expectedResponse = openApiSpec.components.responses[nameOfReferencedResponseDefinition];
+    expectedResponse = findResponseDefinitionObject(expectedResponse['$ref'], openApiSpec);
   }
   if (!expectedResponse) {
     throw new Error(`No '${status}' response defined for endpoint '${httpMethod} ${openApiPath}' in OpenAPI spec`);
   }
 
   return { [status]: expectedResponse };
+}
+
+function findResponseDefinitionObject(referenceString, openApiSpec) {
+  try { // OpenAPI 3
+    const nameOfReferencedResponseDefinition = referenceString.split('#/components/responses/')[1];
+    return openApiSpec.components.responses[nameOfReferencedResponseDefinition];
+  } catch (error) { // OpenAPI 2
+    const nameOfReferencedResponseDefinition = referenceString.split('#/responses/')[1];
+    return openApiSpec.responses[nameOfReferencedResponseDefinition];
+  }
 }
 
 function validateResAgainstApiSpec(res, expectedResponse, openApiSpec) {

--- a/lib/index.js
+++ b/lib/index.js
@@ -34,20 +34,20 @@ module.exports = function (pathToOpenApiSpec) {
     const Assertion = chai.Assertion;
 
     Assertion.addProperty('satisfyApiSpec', function () {
-      const res = new Response(this._obj);
-      const req = res.req();
+      const actualResponse = new Response(this._obj);
+      const actualRequest = actualResponse.req();
 
-      const expectedResponse = findResExpectedByApiSpec(res, openApiSpec);
+      const expectedResponse = findResponseObjDefinedInApiSpec(actualResponse, openApiSpec);
 
-      const validationError = validateResAgainstApiSpec(res, expectedResponse, openApiSpec);
+      const validationError = validateResAgainstApiSpec(actualResponse, expectedResponse, openApiSpec);
       if (validationError) {
-        validationError.actualResponse = res.summary();
+        validationError.actualResponse = actualResponse.summary();
       }
 
       this.assert(
         !validationError,
         `expected res to satisfy API spec:\n${utils.stringify(validationError)}`,
-        `expected res not to satisfy API spec for '${res.status()}' response defined for endpoint '${req.method} ${getOAPath(req, openApiSpec)}' in OpenAPI spec\nres: ${res.toString()}`,
+        `expected res not to satisfy API spec for '${actualResponse.status()}' response defined for endpoint '${actualRequest.method} ${getOAPath(actualRequest, openApiSpec)}' in OpenAPI spec\nres: ${actualResponse.toString()}`,
       );
     });
   };
@@ -82,22 +82,22 @@ function getOAVersion(openApiSpec) {
   return openApiSpec.swagger || openApiSpec.openapi;
 }
 
-function findResExpectedByApiSpec(res, openApiSpec) {
-  const req = res.req();
-  const openApiPath = getOAPath(req, openApiSpec);
-  const pathObj = openApiSpec.paths[openApiPath];
-  if (!pathObj) {
+function findResponseObjDefinedInApiSpec(actualResponse, openApiSpec) {
+  const actualRequest = actualResponse.req();
+  const openApiPath = getOAPath(actualRequest, openApiSpec);
+  const pathItemObject = openApiSpec.paths[openApiPath];
+  if (!pathItemObject) {
     throw new Error(`No '${openApiPath}' path defined in OpenAPI spec`);
   }
 
-  const httpMethod = req.method;
-  const endpointObj = pathObj[httpMethod.toLowerCase()];
-  if (!endpointObj) {
-    throw new Error(`No '${req.method}' method defined for path '${openApiPath}' in OpenAPI spec`);
+  const httpMethod = actualRequest.method;
+  const operationObject = pathItemObject[httpMethod.toLowerCase()];
+  if (!operationObject) {
+    throw new Error(`No '${actualRequest.method}' method defined for path '${openApiPath}' in OpenAPI spec`);
   }
 
-  const status = res.status();
-  let expectedResponse = endpointObj.responses[status];
+  const status = actualResponse.status();
+  let expectedResponse = operationObject.responses[status];
   if (expectedResponse && expectedResponse['$ref']) {
     expectedResponse = findResponseDefinitionObject(expectedResponse['$ref'], openApiSpec);
   }
@@ -118,7 +118,7 @@ function findResponseDefinitionObject(referenceString, openApiSpec) {
   }
 }
 
-function validateResAgainstApiSpec(res, expectedResponse, openApiSpec) {
+function validateResAgainstApiSpec(actualResponse, expectedResponse, openApiSpec) {
   const resValidator = new OpenAPIResponseValidator({
     responses: expectedResponse,
     components: openApiSpec.components, // needed if openApiSpec is OpenAPI 3
@@ -126,19 +126,20 @@ function validateResAgainstApiSpec(res, expectedResponse, openApiSpec) {
   });
 
   const [expectedResStatus] = Object.keys(expectedResponse);
-  const validationError = resValidator.validateResponse(expectedResStatus, res.body());
+  const validationError = resValidator.validateResponse(expectedResStatus, actualResponse.body());
   return validationError;
 }
 
-function getOAPath(req, openApiSpec) {
-  const pathname = url.parse(req.path).pathname; // excludes the query (because: path = pathname + query)
+function getOAPath(actualRequest, openApiSpec) {
+  const pathname = url.parse(actualRequest.path).pathname; // excludes the query (because: path = pathname + query)
 
+  // needed if openApiSpec is OpenAPI 3
   const serverUrl = getServerUrlUsed(pathname, openApiSpec.servers);
-
   // if servers have been defined but none of the servers match the current path
   if (openApiSpec.servers && !serverUrl) {
     throw new Error(`No server matching '${pathname}' path defined in OpenAPI spec`);
   }
+
   const pathnameWithoutServerUrl = removeServerURLFromPath(pathname, serverUrl);
   const OAPath = findOAPathMatchingPathname(pathnameWithoutServerUrl, openApiSpec);
   return OAPath || pathnameWithoutServerUrl;
@@ -155,6 +156,7 @@ function findOAPathMatchingPathname(pathname, openApiSpec) {
   return OAPath;
 }
 
+// for OpenAPI 3
 function getServerUrlUsed(openApiPathWithServerUrl, servers) {
   if (!servers) {
     return null;
@@ -163,6 +165,7 @@ function getServerUrlUsed(openApiPathWithServerUrl, servers) {
   return serverUsed ? serverUsed.url : null;
 }
 
+// for OpenAPI 3
 function removeServerURLFromPath(openApiPathWithServerUrl, serverUrlUsed) {
   return openApiPathWithServerUrl.replace(serverUrlUsed, '');
 }

--- a/test/resources/exampleOpenApiFiles/valid/openapi2.json
+++ b/test/resources/exampleOpenApiFiles/valid/openapi2.json
@@ -134,7 +134,7 @@
         }
       }
     },
-    "/test/responseBody/schemaDef": {
+    "/test/responseBody/referencesSchemaObject": {
       "get": {
         "produces": [
           "application/json"
@@ -142,9 +142,9 @@
         "parameters": [],
         "responses": {
           "200": {
-            "description": "Response body should match a schema",
+            "description": "Response body references a schema object",
             "schema": {
-              "$ref": "#/definitions/ExampleSchema"
+              "$ref": "#/definitions/ExampleSchemaObject"
             }
           }
         }
@@ -191,7 +191,7 @@
     }
   },
   "definitions": {
-    "ExampleSchema": {
+    "ExampleSchemaObject": {
       "type": "string"
     }
   },

--- a/test/resources/exampleOpenApiFiles/valid/openapi2.json
+++ b/test/resources/exampleOpenApiFiles/valid/openapi2.json
@@ -184,10 +184,7 @@
         "parameters": [],
         "responses": {
           "200": {
-            "description": "Response body should be a string",
-            "schema": {
-              "type": "string"
-            }
+            "$ref": "#/responses/Response1"
           }
         }
       }
@@ -198,13 +195,11 @@
       "type": "string"
     }
   },
-  "x-components": {
-    "responses": {
-      "Response1": {
-        "description": "Response body should be a string",
-        "schema": {
-          "type": "string"
-        }
+  "responses": {
+    "Response1": {
+      "description": "Response body should be a string",
+      "schema": {
+        "type": "string"
       }
     }
   }

--- a/test/resources/exampleOpenApiFiles/valid/openapi2.json
+++ b/test/resources/exampleOpenApiFiles/valid/openapi2.json
@@ -6,6 +6,16 @@
     "version": "0.1.0"
   },
   "paths": {
+    "/test/HTTPMethod": {
+      "get": {
+        "parameters": [],
+        "responses": {
+          "204": {
+            "description": "No response body"
+          }
+        }
+      }
+    },
     "/test/multiplePathParams/{param1}/{param2}": {
       "get": {
         "parameters": [
@@ -108,6 +118,22 @@
         }
       }
     },
+    "/test/responseBody/emptyObject": {
+      "get": {
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [],
+        "responses": {
+          "200": {
+            "description": "Response body is an empty object. (Used to test that res.text is populated instead)",
+            "schema": {
+              "type": "object"
+            }
+          }
+        }
+      }
+    },
     "/test/responseBody/schemaDef": {
       "get": {
         "produces": [
@@ -140,7 +166,7 @@
         }
       }
     },
-    "/test/HTTPMethod": {
+    "/test/responseStatus": {
       "get": {
         "parameters": [],
         "responses": {
@@ -150,12 +176,18 @@
         }
       }
     },
-    "/test/responseStatus": {
+    "/test/resReferencesResDefinition": {
       "get": {
+        "produces": [
+          "application/json"
+        ],
         "parameters": [],
         "responses": {
-          "204": {
-            "description": "No response body"
+          "200": {
+            "description": "Response body should be a string",
+            "schema": {
+              "type": "string"
+            }
           }
         }
       }
@@ -166,5 +198,14 @@
       "type": "string"
     }
   },
-  "x-components": {}
+  "x-components": {
+    "responses": {
+      "Response1": {
+        "description": "Response body should be a string",
+        "schema": {
+          "type": "string"
+        }
+      }
+    }
+  }
 }

--- a/test/resources/exampleOpenApiFiles/valid/openapi2.json
+++ b/test/resources/exampleOpenApiFiles/valid/openapi2.json
@@ -176,7 +176,7 @@
         }
       }
     },
-    "/test/resReferencesResDefinition": {
+    "/test/responseReferencesResponseDefinitionObject": {
       "get": {
         "produces": [
           "application/json"
@@ -184,7 +184,7 @@
         "parameters": [],
         "responses": {
           "200": {
-            "$ref": "#/responses/Response1"
+            "$ref": "#/responses/ExampleResponseDefinitionObject"
           }
         }
       }
@@ -196,7 +196,7 @@
     }
   },
   "responses": {
-    "Response1": {
+    "ExampleResponseDefinitionObject": {
       "description": "Response body should be a string",
       "schema": {
         "type": "string"

--- a/test/resources/exampleOpenApiFiles/valid/openapi3.yml
+++ b/test/resources/exampleOpenApiFiles/valid/openapi3.yml
@@ -41,6 +41,11 @@ paths:
             application/json:
               schema:
                 type: object
+  /test/resReferencesResDefinition:
+    get:
+      responses:
+        200:
+          $ref: '#/components/responses/Response1'
   /test/multipleResponsesDefined:
     get:
       responses:
@@ -115,3 +120,10 @@ components:
   schemas:
     ExampleSchema:
       type: string
+  responses:
+    Response1:
+      description: Response body should be a string
+      content:
+        application/json:
+          schema:
+            type: string

--- a/test/resources/exampleOpenApiFiles/valid/openapi3.yml
+++ b/test/resources/exampleOpenApiFiles/valid/openapi3.yml
@@ -18,15 +18,15 @@ paths:
             application/json:
               schema:
                 type: string
-  /test/responseBody/schemaDef:
+  /test/responseBody/referencesSchemaObject:
     get:
       responses:
         200:
-          description: Response body should match a schema
+          description: Response body references a schema object
           content:
             application/json:
               schema:
-                $ref: '#/components/schemas/ExampleSchema'
+                $ref: '#/components/schemas/ExampleSchemaObject'
   /test/responseBody/empty:
     get:
       responses:
@@ -118,7 +118,7 @@ paths:
           description: No response body
 components:
   schemas:
-    ExampleSchema:
+    ExampleSchemaObject:
       type: string
   responses:
     Response1:

--- a/test/resources/exampleOpenApiFiles/valid/openapi3.yml
+++ b/test/resources/exampleOpenApiFiles/valid/openapi3.yml
@@ -41,11 +41,11 @@ paths:
             application/json:
               schema:
                 type: object
-  /test/resReferencesResDefinition:
+  /test/responseReferencesResponseDefinitionObject:
     get:
       responses:
         200:
-          $ref: '#/components/responses/Response1'
+          $ref: '#/components/responses/ExampleResponseDefinitionObject'
   /test/multipleResponsesDefined:
     get:
       responses:
@@ -121,7 +121,7 @@ components:
     ExampleSchemaObject:
       type: string
   responses:
-    Response1:
+    ExampleResponseDefinitionObject:
       description: Response body should be a string
       content:
         application/json:

--- a/test/unit/assertions/definedServerPath.test.js
+++ b/test/unit/assertions/definedServerPath.test.js
@@ -30,7 +30,7 @@ describe('Using an OA3 spec that defines server paths', function () {
       status: 200,
       req: {
         method: 'GET',
-        path: `${differentServer}/test/responseBody/schemaDef`,
+        path: `${differentServer}/test/responseBody/string`,
       },
       body: 'valid body (string)',
     };
@@ -50,19 +50,19 @@ describe('Using an OA3 spec that defines server paths', function () {
       status: 200,
       req: {
         method: 'GET',
-        path: `${differentServer}/test/responseBody/schemaDef`,
+        path: `${differentServer}/test/responseBody/string`,
       },
       body: 'valid body (string)',
     };
 
     it('fails', function () {
       const assertion = () => expect(res).to.satisfyApiSpec;
-      expect(assertion).to.throw('No server matching \'/missing/test/responseBody/schemaDef\' path defined in OpenAPI spec');
+      expect(assertion).to.throw('No server matching \'/missing/test/responseBody/string\' path defined in OpenAPI spec');
     });
 
     it('fails when using .not', function () {
       const assertion = () => expect(res).to.not.satisfyApiSpec;
-      expect(assertion).to.throw('No server matching \'/missing/test/responseBody/schemaDef\' path defined in OpenAPI spec');
+      expect(assertion).to.throw('No server matching \'/missing/test/responseBody/string\' path defined in OpenAPI spec');
     });
   });
 });

--- a/test/unit/assertions/satisfyApiSpec.test.js
+++ b/test/unit/assertions/satisfyApiSpec.test.js
@@ -112,8 +112,35 @@ for (const spec of openApiSpecs) {
           });
         });
 
-        describe('res.req.path matches multiple expected responses', function () {
-          describe('1st expected response', function () {
+        describe('res.req.path matches a response referencing a response definition', function () {
+          const res = {
+            status: 200,
+            req: {
+              method: 'GET',
+              path: `${serverPath}/test/resReferencesResDefinition`,
+            },
+            body: 'valid body (string)',
+          };
+
+          it('passes', function () {
+            expect(res).to.satisfyApiSpec;
+          });
+
+          it('fails when using .not', function () {
+            const assertion = () => expect(res).to.not.satisfyApiSpec;
+            expect(assertion).to.throw(
+              `expected res not to satisfy API spec for '200' response defined for endpoint 'GET /test/resReferencesResDefinition' in OpenAPI spec\nres: ${
+                util.inspect({
+                  status: 200,
+                  body: 'valid body (string)',
+                })
+              }`
+            );
+          });
+        });
+
+        describe('res.req.path matches a path with multiple defined responses', function () {
+          describe('1st defined response', function () {
             const res = {
               status: 201,
               req: {
@@ -133,7 +160,7 @@ for (const spec of openApiSpecs) {
             });
           });
 
-          describe('2nd expected response', function () {
+          describe('2nd defined response', function () {
             const res = {
               status: 202,
               req: {
@@ -153,7 +180,7 @@ for (const spec of openApiSpecs) {
             });
           });
 
-          describe('3rd expected response', function () {
+          describe('3rd defined response', function () {
             const res = {
               status: 203,
               req: {

--- a/test/unit/assertions/satisfyApiSpec.test.js
+++ b/test/unit/assertions/satisfyApiSpec.test.js
@@ -72,12 +72,12 @@ for (const spec of openApiSpecs) {
             });
           });
 
-          describe('match a (string) schema', function () {
+          describe('match a (string) schema object', function () {
             const res = {
               status: 200,
               req: {
                 method: 'GET',
-                path: `${serverPath}/test/responseBody/schemaDef`,
+                path: `${serverPath}/test/responseBody/referencesSchemaObject`,
               },
               body: 'valid body (string)',
             };
@@ -88,7 +88,7 @@ for (const spec of openApiSpecs) {
 
             it('fails when using .not', function () {
               const assertion = () => expect(res).to.not.satisfyApiSpec;
-              expect(assertion).to.throw('expected res not to satisfy API spec for \'200\' response defined for endpoint \'GET /test/responseBody/schemaDef\' in OpenAPI spec');
+              expect(assertion).to.throw('expected res not to satisfy API spec for \'200\' response defined for endpoint \'GET /test/responseBody/referencesSchemaObject\' in OpenAPI spec');
             });
           });
 

--- a/test/unit/assertions/satisfyApiSpec.test.js
+++ b/test/unit/assertions/satisfyApiSpec.test.js
@@ -117,7 +117,7 @@ for (const spec of openApiSpecs) {
             status: 200,
             req: {
               method: 'GET',
-              path: `${serverPath}/test/resReferencesResDefinition`,
+              path: `${serverPath}/test/responseReferencesResponseDefinitionObject`,
             },
             body: 'valid body (string)',
           };
@@ -129,7 +129,7 @@ for (const spec of openApiSpecs) {
           it('fails when using .not', function () {
             const assertion = () => expect(res).to.not.satisfyApiSpec;
             expect(assertion).to.throw(
-              `expected res not to satisfy API spec for '200' response defined for endpoint 'GET /test/resReferencesResDefinition' in OpenAPI spec\nres: ${
+              `expected res not to satisfy API spec for '200' response defined for endpoint 'GET /test/responseReferencesResponseDefinitionObject' in OpenAPI spec\nres: ${
                 util.inspect({
                   status: 200,
                   body: 'valid body (string)',

